File: app/src/main/java/com/commit451/commitspiration/activity/MainActivity.java
package com.commit451.commitspiration.activity;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.commit451.commitspiration.R;
import com.commit451.commitspiration.api.WhatTheCommitClient;
import com.commit451.commitspiration.model.WhatTheCommitData;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import timber.log.Timber;

public class MainActivity extends AppCompatActivity {

    Toolbar mToolbar;
    ViewGroup mRoot;
    TextView mMessage;

    WhatTheCommitData mWhatTheCommitData;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mToolbar = (Toolbar) findViewById(R.id.toolbar);
        mRoot = (ViewGroup) findViewById(R.id.root);
        mMessage = (TextView) findViewById(R.id.message);

        mToolbar.setTitle(R.string.app_name);
        mToolbar.inflateMenu(R.menu.main);
        mToolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.action_share:
                        if (mWhatTheCommitData != null) {
                            Intent sendIntent = new Intent();
                            sendIntent.setAction(Intent.ACTION_SEND);
                            sendIntent.putExtra(Intent.EXTRA_TEXT, mWhatTheCommitData.url);
                            sendIntent.setType("text/plain");
                            startActivity(sendIntent);
                        } else {
                            Snackbar.make(mRoot, "Nothing to share yet!", Snackbar.LENGTH_LONG)
                                    .show();
                        }
                        return true;
                    case R.id.action_refresh:
                        load();
                        return true;
                }
                return false;
            }
        });
        mMessage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mWhatTheCommitData != null) {
                    // Gets a handle to the clipboard service.
                    ClipboardManager clipboard = (ClipboardManager)
                            getSystemService(Context.CLIPBOARD_SERVICE);
                    // Creates a new text clip to put on the clipboard
                    ClipData clip = ClipData.newPlainText(mWhatTheCommitData.url, mWhatTheCommitData.message);
                    clipboard.setPrimaryClip(clip);
                    Snackbar.make(mRoot, "Copied to clipboard", Snackbar.LENGTH_SHORT)
                            .show();
                } else {
                    Snackbar.make(mRoot, "Nothing to copy yet", Snackbar.LENGTH_SHORT)
                            .show();
                }
            }
        });
        load();
    }

    private void load() {
        WhatTheCommitClient.instance().getMessage(WhatTheCommitClient.API_URL).enqueue(new Callback<WhatTheCommitData>() {
            @Override
            public void onResponse(Call<WhatTheCommitData> call, Response<WhatTheCommitData> response) {
                if (response.isSuccessful()) {
                    mWhatTheCommitData = response.body();
                    mMessage.setText(mWhatTheCommitData.message);
                } else {
                    showError();
                }
            }

            @Override
            public void onFailure(Call<WhatTheCommitData> call, Throwable t) {
                Timber.e(t, null);
                showError();
            }
        });
    }

    private void showError() {
        Snackbar.make(mRoot, "Unable to get that commit message for ya, sorry", Snackbar.LENGTH_LONG)
                .show();
    }
}

File: app/src/main/java/com/commit451/commitspiration/activity/MainActivity.kt
package com.commit451.commitspiration.activity

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import com.commit451.commitspiration.R
import com.commit451.commitspiration.api.WhatTheCommitClient
import com.commit451.commitspiration.extensions.snack
import com.commit451.commitspiration.model.WhatTheCommitData
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber

class MainActivity : AppCompatActivity() {

    @BindView(R.id.root) lateinit var root: ViewGroup
    @BindView(R.id.toolbar) lateinit var toolbar: Toolbar
    @BindView(R.id.text_message) lateinit var textMessage: TextView
    @BindView(R.id.progress) lateinit var progress: View

    var whatTheCommitData: WhatTheCommitData? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)

        toolbar.setTitle(R.string.app_name)
        toolbar.inflateMenu(R.menu.main)
        toolbar.setOnMenuItemClickListener(Toolbar.OnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_share -> {
                    if (whatTheCommitData != null) {
                        val sendIntent = Intent()
                        sendIntent.action = Intent.ACTION_SEND
                        sendIntent.putExtra(Intent.EXTRA_TEXT, whatTheCommitData!!.url)
                        sendIntent.type = "text/plain"
                        startActivity(sendIntent)
                    } else {
                        root.snack("Nothing to share yet!")
                    }
                    return@OnMenuItemClickListener true
                }
                R.id.action_copy -> {
                    if (whatTheCommitData != null) {
                        // Gets a handle to the clipboard service.
                        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                        // Creates a new text clip to put on the clipboard
                        val clip = ClipData.newPlainText(whatTheCommitData!!.url, whatTheCommitData!!.message)
                        clipboard.primaryClip = clip

                        root.snack("Copied to clipboard")
                    } else {
                        root.snack("Nothing to copy yet")
                    }
                    return@OnMenuItemClickListener true
                }
            }
            false
        })
        root.setOnClickListener {
            load()
        }
        load()
    }

    private fun load() {
        showLoading()
        WhatTheCommitClient.service.getMessage(WhatTheCommitClient.API_URL).enqueue(object : Callback<WhatTheCommitData> {
            override fun onResponse(call: Call<WhatTheCommitData>, response: Response<WhatTheCommitData>) {
                hideLoading()
                if (response.isSuccessful) {
                    whatTheCommitData = response.body()
                    textMessage.text = whatTheCommitData!!.message
                } else {
                    showError()
                }
            }

            override fun onFailure(call: Call<WhatTheCommitData>, t: Throwable) {
                Timber.e(t)
                hideLoading()
                showError()
            }
        })
    }

    private fun showLoading() {
        textMessage.animate()?.alpha(0.0f)
        progress.animate()?.alpha(1.0f)
    }

    private fun hideLoading() {
        textMessage.animate()?.alpha(1.0f)
        progress.animate()?.alpha(0.0f)
    }

    private fun showError() {
        root.snack("Unable to get that commit message for ya, sorry")
    }
}
